app.py

```diff
@@ -1,12 +1,115 @@
 import streamlit as st
+import streamlit.components.v1 as components
 from pathlib import Path
 import json
-# For some reason the windows version only works if this is imported here
-import pyopenms
+import os
+import time
+import threading
+import signal
+import tornado.web
+import tornado.ioloop
+import pyopenms  # required import for Windows
 
+# Keep settings at the very top
 if "settings" not in st.session_state:
-        with open("settings.json", "r") as f:
-            st.session_state.settings = json.load(f)
+    with open("settings.json", "r") as f:
+        st.session_state.settings = json.load(f)
+
+# --- Global Session Counter ---
+global_active_sessions = 0
+last_heartbeat = 0   # Global heartbeat timestamp
+
+def increment_active_sessions():
+    global global_active_sessions
+    global_active_sessions += 1
+    print(f"Session connected, total active sessions: {global_active_sessions}")
+
+def decrement_active_sessions():
+    global global_active_sessions
+    global_active_sessions -= 1
+    print(f"Session disconnected, total active sessions: {global_active_sessions}")
+    if global_active_sessions <= 0:
+        shutdown()
+
+def shutdown():
+    print("🚨 No active users. Shutting down Streamlit server...")
+    os.kill(os.getpid(), signal.SIGTERM)
+
+# Register this session only once.
+if "registered" not in st.session_state:
+    st.session_state.registered = True
+    increment_active_sessions()
+
+# --- Tornado Endpoints ---
+
+# This handler is called on browser unload (if it fires)
+class CloseAppHandler(tornado.web.RequestHandler):
+    def post(self):
+        time.sleep(1)
+        decrement_active_sessions()
+        self.write("OK")
+
+# This handler receives heartbeat pings
+class HeartbeatHandler(tornado.web.RequestHandler):
+    def post(self):
+        global last_heartbeat
+        last_heartbeat = time.time()
+        self.write("OK")
+
+def start_tornado_server():
+    routes = [
+        (r"/_closeapp", CloseAppHandler),
+        (r"/heartbeat", HeartbeatHandler)
+    ]
+    app = tornado.web.Application(routes)
+    app.listen(port=8501)  # Adjust if needed; matches Streamlit port.
```

Comment:
[should be same port in .toml file](https://github.com/OpenMS/streamlit-template/blob/main/.streamlit/config.toml)